### DISCUSSION TITLE
where is the login page chinese word change?

### DISCUSSION THREAD

**Discussion by XR-stb** [FreeRDP/FreeRDP#9158] [Q&A] [answered]
like the title, i can find the english word position and to change, but the chinese word i can't find.
i want to know the freerdp gdi logic, can anyone help me, very thanks
![image](https://github.com/FreeRDP/FreeRDP/assets/56108982/bb14c60a-cfc8-4a2c-b6e9-3a0df173ae99)

**Comment by hardening** [comment]
Do you mean that the "Enter your credentials" label is drawn using glyphs and that you don't find the equivalent for Chinese labels ?

**Reply by XR-stb** [reply]
yes，i can't find '用户名' string to modify，and can you tall me the ui how to render, very thank your

**Reply by akallabeth** [reply, marked as answer]
@XR-stb you can not. `wfreerdp` (which you use, correct?) just calls `CredUIPromptForCredentialsW` which provides these translations.

**Reply by hardening** [reply]
Well, I imagine that the server draws Chinese labels using some kind of bitmap update depending on the enabled codecs...

**Reply by XR-stb** [reply]
Well, I have one last question, what is the main complex function of Windows GDI in FreeRDP, I have seen a lot of code, but as far as I see, it seems to be just rewriting the Windows GDI API

**Reply by akallabeth** [reply]
@XR-stb what do you mean?
there are some parts (for the old `ROP` style `RDP` protocol from `XP` and `NT` era) that need to recrate that (`FreeRDP` is used on many platforms), but everything more modern uses different drawing paths.

**Reply by XR-stb** [reply]
None of these plotting functions seem to be called, I would like to know where they are used, can you tell me, thank you very much
![image](https://github.com/FreeRDP/FreeRDP/assets/56108982/0bfdc394-7b3a-4f29-ba13-76dc56af96b8)

**Reply by lazy5f** [reply]
I was also one of those who were wondering about when and how they are called. I'm not sure there is some relavent documentation, but if I share my experience, I started by exploring "gdi" command-line option. (You can see the function is called if `!settings->SoftwareGdi`.)

**Reply by akallabeth** [reply]
@XR-stb they are implementations for the ancient `WinXP/NT` aera graphics primitives redirection. `FreeRDP` provides a default software based implementation in the core library and some clients override these with some hardware based implementation (`FreeRDP_SoftwareGdi` is the setting for that)
that being said, it is obsolete nowadays as the UI consists mostly only of bitmaps anymore, so they are mostly useless and slow (newer systems prefer other drawing modes like the `SurfaceBits` of `win7` or the `rdpgfx` style redirection starting with `windows 8`)
